[CON-1666] feat(insightly): Metadata custom objects & fields

Cobalt0s opened the pull request:
# Description

* Core object metadata still comes from `schema.json` since there’s no public API for it. The internal Insightly Dashboard has access, but we don’t.is privilige, therefore it is internal to them.
* The `CustomFields` endpoint provides everything needed to populate `common.FieldMetadata`.
* Objects ending with `__c` endpoint provides everything needed to populate `CustomObjects` endpoint.
* Custom objects support standard CRUD operations (GET/POST/PUT/DELETE). The response format is captured by `customObjectSchema` variable. Their responses follow the `customObjectSchema`, with fields nested under `CUSTOMFIELDS`. Metadata for those fields is returned at the root level—just like core objects.

# Result
* If a custom object like `fruits` has a `color` field, it appears in metadata and is readable via the connector.
* Core objects (e.g., `contact`) can also have custom fields (e.g., `interests`). These are included in metadata and can be queried the same way by the read.

# Tests
## Contacts
![image](https://github.com/user-attachments/assets/1d39bf96-57d7-4149-8a23-ee8544aaeba2)
![image](https://github.com/user-attachments/assets/bab0578f-07b7-4636-81a2-eb0623452367)

## Fruits the Custom Object
![image](https://github.com/user-attachments/assets/293a7cbf-888a-4f3b-b523-6b72d0ef5ea1)

Cobalt0s commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/amp-labs/connectors/1589?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#1589** <a href="https://app.graphite.dev/github/pr/amp-labs/connectors/1589?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/amp-labs/connectors/1589?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#1587** <a href="https://app.graphite.dev/github/pr/amp-labs/connectors/1587?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#1579** <a href="https://app.graphite.dev/github/pr/amp-labs/connectors/1579?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#1578** <a href="https://app.graphite.dev/github/pr/amp-labs/connectors/1578?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#1577** <a href="https://app.graphite.dev/github/pr/amp-labs/connectors/1577?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#1576** <a href="https://app.graphite.dev/github/pr/amp-labs/connectors/1576?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#1575** <a href="https://app.graphite.dev/github/pr/amp-labs/connectors/1575?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#1560** <a href="https://app.graphite.dev/github/pr/amp-labs/connectors/1560?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

jlimatampersand reviewed (changes requested):
Logic makes seems good. Some minor comments left.

Cobalt0s commented on providers/insightly/custom.go:
```diff
@@ -76,3 +80,150 @@ func flattenCustomFields() common.RecordTransformer {
+type customFieldsResponse []customFieldResponse
+
+// nolint:tagliatelle
+type customFieldResponse struct {
```
Good point. This indeed needs documentation

Cobalt0s commented on providers/insightly/custom.go:
```diff
@@ -76,3 +80,150 @@ func flattenCustomFields() common.RecordTransformer {
+func (c *Connector) requestCustomFields(
+	ctx context.Context, objectName string,
+) (customFieldsRegistry, error) {
+	url, err := c.getCustomFieldsURL(objectName)
```
I think the description I have above is a bit misleading. While the API returns 4xx on empty object and thats the only way to get a failing response the deep connector never passes empty objectName.

I think to avoid assumptions that we have I will keep the description and return an error for empty objectName so the method is self-sustained.

Cobalt0s replied:
I documented all possible API behaviours during research/exploration stage of development.

Cobalt0s commented on providers/insightly/metadata.go:
```diff
@@ -0,0 +1,111 @@
+}
+
+func (c *Connector) getObjectMetadata(ctx context.Context, objectName string) (*common.ObjectMetadata, error) {
+	objectData, err := metadata.Schemas.SelectOne(c.Module(), objectName)
```
SchemaProvider is out of the box simple implementation of metadata coming from static file. There are some other objects like this in the `components` package. But none of them apply here because this one is a very particular case.

Insightly should still serve metadata from static file with one caveat which if the object has custom suffix then we know the object can be described by making an API call. So in practice this one is a mixture of `SchemaProvider` + `Custom Object API call`.

Cobalt0s commented on providers/insightly/metadata.go:
```diff
@@ -0,0 +1,111 @@
+	objectData, err := metadata.Schemas.SelectOne(c.Module(), objectName)
+	if err != nil {
+		// Check if the object is custom, then we still can create metadata.
+		if errors.Is(err, common.ErrObjectNotSupported) && strings.HasSuffix(objectName, customMarker) {
```
Yeah, that sounds more logical with better intent.

Cobalt0s commented on providers/insightly/metadata.go:
```diff
@@ -0,0 +1,111 @@
+	}
+}
+
+func getDefaultValues(field customFieldResponse) common.FieldValues {
```
Yes. getFieldValues would be better. In comments I might use enums but since we called it FieldValues in common I will stick to that.

Cobalt0s replied:
You know, I decided to make it a member function on this type:
```
func (c customFieldResponse) getValues() common.FieldValues {
```
Usage:
```
Values:       field.getValues(),
```
This method is not colliding with other functions and should make more sense that it returns `FieldValues`, it fits nicely I think.

Cobalt0s commented on providers/insightly/custom.go:
```diff
@@ -76,3 +80,150 @@ func flattenCustomFields() common.RecordTransformer {
+}
+
+func (c customFieldResponse) BelongsToObject(objectName string) bool {
+	if c.FieldFor == objectName {
```
Turns out customSuffix could be in caps `__C` and `__c`. It was caught in the mock test adapted code a bit.

Cobalt0s commented:
@jlimatampersand 
Addressed comments in the following commit: https://github.com/amp-labs/connectors/pull/1589/commits/731709184fc90830af8dad078cb2633c94002c59

jlimatampersand reviewed (approved):
Sorry I totally lost track of this. I will keep my eye on PRs more closely.
